src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final double SIZE_INCREASE_FACTOR = 1.5;
+    private Object[] elementData; // should we use transient keyword?
+    private int size;
+
+    public ArrayList() {
+        elementData = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = buildUp();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndex(index);
+        if (size == elementData.length) {
+            elementData = buildUp();
+        }
+        System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        checkList(list);
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        elementData[index] = value;
     }
 
     @Override
+    @SuppressWarnings("unchecked")
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        checkIndexEqualsSize(index);
+        final T oldValue = (T) elementData[index];
+        if (size - 1 > index) {
+            System.arraycopy(elementData, index + 1, elementData, index, size - index - 1);
+        }
+        elementData[size - 1] = null;
+        size--;
+        return oldValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int index = getElement(element);
+        if (index == -1) {
+            throw new NoSuchElementException("This element does not exist: " + element);
+        }
+        remove(index);
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Object[] buildUp() {
```

Comment:
```suggestion
    private void resize() {
```

could we make this method void?